### PR TITLE
Random correlated OT traits + ideal impls

### PR DESCRIPTION
**Pull request by sinui0**
This PR adds Random COT traits and ideal implementations.

@xiangxiecrypto reminded me that Ferret implements these, not the chosen variants.

### CODE REVIEW

**Comment by xiangxiecrypto** [comment]
You can implement this trait for KOS15 (or other protocols), which is also a building block of Ferret.

File: ot/mpz-ot/src/lib.rs [hunk 1]
```diff
@@ -92,6 +92,32 @@ where
     ) -> Result<(), OTError>;
 }
 
+/// A random correlated oblivious transfer sender.
+#[async_trait]
+pub trait RandomCOTSender<T>: ProtocolMessage
+where
+    T: Send + Sync,
+{
+    /// Obliviously transfers the correlations to the receiver.
```
Comment (themighty1) [inline ot/mpz-ot/src/lib.rs]: I changed "correlation" to "correlated message" for consistency and greater clarity here and in the rest of my suggestions.
```suggestion
    /// Obliviously transfers the correlated messages to the receiver.
```

File: ot/mpz-ot/src/lib.rs [hunk 2]
```diff
@@ -92,6 +92,32 @@ where
...
+    ///
+    /// Returns the `0`-bit messages that were obliviously transferred.
+    ///
+    /// # Arguments
+    ///
+    /// * `sink` - The IO sink to the receiver.
+    /// * `stream` - The IO stream from the receiver.
+    /// * `count` - The number of correlations to obliviously transfer.
```
Comment (themighty1) [inline ot/mpz-ot/src/lib.rs]:
```suggestion
    /// * `count` - The number of correlated messages to obliviously transfer.
```

File: ot/mpz-ot/src/lib.rs [hunk 3]
```diff
@@ -139,6 +165,33 @@ where
...
+{
+    /// Obliviously receives correlated messages with random choices.
+    ///
+    /// Returns a tuple of the choices and the messages, respectively.
+    ///
+    /// # Arguments
+    ///
+    /// * `sink` - The IO sink to the sender.
+    /// * `stream` - The IO stream from the sender.
+    /// * `count` - The number of correlations to obliviously receive.
```
Comment (themighty1) [inline ot/mpz-ot/src/lib.rs]:
```suggestion
    /// * `count` - The number of correlated messages to obliviously receive.
```

File: ot/mpz-ot/src/lib.rs [hunk 4]
```diff
@@ -266,6 +319,20 @@ pub trait COTSenderShared<T> {
...
+#[async_trait]
+pub trait RandomCOTSenderShared<T> {
+    /// Obliviously transfers correlated messages to the receiver.
+    ///
+    /// Returns the `0`-bit messages that were obliviously transferred.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - The unique identifier for this transfer.
+    /// * `count` - The number of correlations to obliviously transfer.
```
Comment (themighty1) [inline ot/mpz-ot/src/lib.rs]:
```suggestion
    /// * `count` - The number of correlated messages to obliviously transfer.
```

File: ot/mpz-ot/src/lib.rs [hunk 5]
```diff
@@ -290,6 +357,24 @@ pub trait COTReceiverShared<T, U> {
...
+#[async_trait]
+pub trait RandomCOTReceiverShared<T, U> {
+    /// Obliviously receives correlated messages with random choices.
+    ///
+    /// Returns a tuple of the choices and the messages, respectively.
+    ///
+    /// # Arguments
+    ///
+    /// * `id` - The unique identifier for this transfer.
+    /// * `count` - The number of correlations to obliviously receive.
```
Comment (themighty1) [inline ot/mpz-ot/src/lib.rs]:
```suggestion
    /// * `count` - The number of correlated messages to obliviously receive.
```

**Review by themighty1 — APPROVED** [review]
ack

**Review by th4s — APPROVED** [review]
:+1:

**Pull request merged**